source/agora/consensus/protocol/Nominator.d

```diff
@@ -451,8 +429,38 @@ extern(D):
             return;  // slot was already externalized, ignore outdated message
         }
 
-        if (this.scp.receiveEnvelope(envelope) != SCP.EnvelopeState.VALID)
+        if (this.scp.receiveEnvelope(envelope) == SCP.EnvelopeState.VALID)
+        {
+            vector!SCPEnvelope envelopes =
+                this.scp.getCurrentState(envelope.statement.slotIndex);
```

Comment:
I thought about this more. I think it's inefficient if we do this every time we receive an envelope.

Instead we should do it:
- On node shutdown
- On node crash

For the first part I suggest adding an override of `shutdown` in the `Validator` class:

```
override void shutdown { }
{
    super.shutdown();
    this.nominator.storeLatestState();
}
```

And implementing `storeLatestState` which will store the envelopes to the database.

For node crashes we actually need to implement a crash / segfault event handler, however this is a separate problem and should not be solved as part of #1170.

Reply:
Yes, that's right.
I thought it was inefficient when I saw the frequency of calls.